bin/faucet/src/server/pow.rs

```diff
@@ -1,16 +1,101 @@
-use std::time::{SystemTime, UNIX_EPOCH};
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+    time::{SystemTime, UNIX_EPOCH},
+};
 
 use axum::{Json, extract::State, response::IntoResponse};
 use miden_tx::utils::ToHex;
 use rand::{Rng, rng};
 use serde::Serialize;
 use sha3::{Digest, Sha3_256};
+use tokio::time::{Duration, interval};
 
 use super::{Server, get_tokens::InvalidRequest};
 
 const DIFFICULTY: u64 = 5;
 const SERVER_TIMESTAMP_TOLERANCE_SECONDS: u64 = 30;
 
+#[derive(Clone)]
+pub struct Challenge {
+    timestamp: u64,
+    server_signature: String,
+}
+
+#[derive(Clone)]
+pub struct ChallengeState {
+    challenges: Arc<Mutex<HashMap<String, Challenge>>>,
+}
+
+impl ChallengeState {
+    pub fn new() -> Self {
+        Self {
+            challenges: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    pub fn add_challenge(&self, seed: &str, server_signature: String, timestamp: u64) {
+        let mut challenges = self.challenges.lock().unwrap();
+        challenges.insert(seed.to_string(), Challenge { timestamp, server_signature });
+    }
+
+    pub fn validate_challenge(
+        &self,
+        seed: &str,
+        server_signature: &str,
+    ) -> Result<(), InvalidRequest> {
+        let challenges = self.challenges.lock().unwrap();
+
+        if let Some(challenge) = challenges.get(seed) {
+            if challenge.server_signature != server_signature {
+                return Err(InvalidRequest::ServerSignaturesDoNotMatch);
+            }
+
+            let current_time = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .expect("Time went backwards")
+                .as_secs();
+
+            if (current_time - challenge.timestamp) > SERVER_TIMESTAMP_TOLERANCE_SECONDS {
+                return Err(InvalidRequest::ExpiredServerTimestamp(
+                    challenge.timestamp,
+                    current_time,
+                ));
+            }
+
+            Ok(())
+        } else {
+            Err(InvalidRequest::InvalidChallenge)
+        }
+    }
+
+    pub fn cleanup_expired_challenges(&self) {
+        let current_time = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Time went backwards")
+            .as_secs();
+
+        let mut challenges = self.challenges.lock().unwrap();
+        challenges.retain(|_, challenge| {
+            (current_time - challenge.timestamp) <= SERVER_TIMESTAMP_TOLERANCE_SECONDS
+        });
+    }
+
+    pub fn remove_challenge(&self, seed: &str) {
+        let mut challenges = self.challenges.lock().unwrap();
+        challenges.remove(seed);
+    }
+}
+
+pub async fn start_cleanup_task(challenge_state: ChallengeState) {
```

Comment:
Do we need timestamp-based cleanup at all? Couldn't we do the LRU cache approach that @igamigo was doing in another PR? That way we don't need a cleanup task either? We just drop the oldest challenge when we are at capacity and adding a new one?

Then again that would create an opportunity for attack if users could create enough challenges to reach our limit continuously. Maybe easy to prevent / control but still.

Just wondering. No change required if you prefer time based.

Reply:
As you mentioned, using an LRU cache allows attackers to keep it full and would cause that valid challenges would be rejected.

Reply:
This also goes back to the comment above where I realized the impl was different to my assumptions 👍